### ISSUE TITLE
KeyError: xxxxxxxxxx when calling optimizer.state_dict()

### ISSUE THREAD

**Issue by AhmadM-DL** [pytorch/pytorch#42428]
## 🐛 Bug

When calling optimizer.state_dict() it raises a KeyError:

```
/root/module/class.py in save_model_parameters(self, model_parameters_path, epoch, optimizer)
    254                       'state_dict': self.state_dict()}
    255         if optimizer:
--> 256             model_dict['optimizer'] = optimizer.state_dict()
    257 
    258         torch.save(model_dict, model_parameters_path)

/usr/local/lib/python3.6/dist-packages/torch/optim/optimizer.py in state_dict(self)
     96         # Remap state to use order indices as keys
     97         packed_state = {(param_mappings[id(k)] if isinstance(k, torch.Tensor) else k): v
---> 98                         for k, v in self.state.items()}
     99         return {
    100             'state': packed_state,

/usr/local/lib/python3.6/dist-packages/torch/optim/optimizer.py in <dictcomp>(.0)
     96         # Remap state to use order indices as keys
     97         packed_state = {(param_mappings[id(k)] if isinstance(k, torch.Tensor) else k): v
---> 98                         for k, v in self.state.items()}
     99         return {
    100             'state': packed_state,

KeyError: 140413627951216
```

## To Reproduce

It is not straightforward to reproduce the error as it doesn't happen in all environments.
It runs normally locally but it raises an error on colab or cloud.

I am training a deep convolution model unsupervised-ly based on [deepcluster ](https://github.com/facebookresearch/deepcluster) pipeline. The approach alternates as follows:

0- remove classifier module
1- feed inputs
2- compute features
3- cluster inputs based on k means
4- add new classifier module
5- use inputs cluster assignments as pseudo labels to train the model

Since I am adding and removing classifier module on the go, I am using `optimizer.add_param_group()` to add new optimization parameters for the added classifier module and using `del optimizer.param_groups[1]` to delete the parameters when the module is removed.

## Expected behavior

The call should not lead to an error.

**Comment by AhmadM-DL** [comment]
It seems that the problem resides in the following part of `optimizer.state_dict()` method:
`param_mappings[id(k)]`. The id of a certain parameter in the `optimizer.state` doesn’t exist in `param_mappings`.

Do you have any thoughs?

I believe that the problem resides in the fact that I am removing `param_groups` as following:
`del optimizer.param_groups[1]`

This is removing the second param group from the optimizer but it seems that it’s associated parameters in `optimizer.state` reside leading to an error when matching `param_mappings` with `optimizer.state`.

If that is the case then what is the right way to remove param_groups? (no documentation regarding this in PyTorch optimizer documentation)

**Comment by mariosasko** [comment]
@AhmadM-DL yes, you are right about the cause of the issue. Each `optim.step` call stores/rewrites in the `optimizer.state` optimizer-specific information (exponential average, etc.) about model parameters (these parameters are the keys) so even if you remove a `param_group`, its parameters are still stored in the `optimizer.state` (the next `optim.step` call doesn't touch these parameters and only information about the parameters from other `param_groups` gets upated).

This line from you bothers me: It runs normally locally but it raises an error on colab or cloud.
It should be possible to reproduce the error regardless of the environment.

@agolynski After triage review, maybe I can take this issue and make a PR if there is something to be fixed.

**Comment by AhmadM-DL** [comment]
> This line from you bothers me: It runs normally locally but it raises an error on colab or cloud.
It should be possible to reproduce the error regardless of the environment.

Yup, that is annoying. Though it is actually happening.
I can privately send you a link to a colab notebook that raises the error.

**Comment by vincentqb** [comment]
> Since I am adding and removing classifier module on the go, I am using optimizer.add_param_group() to add new optimization parameters for the added classifier module and using del optimizer.param_groups[1] to delete the parameters when the module is removed.

Do you have a minimal example that reproduces the issue?

**Comment by AhmadM-DL** [comment]
> > Since I am adding and removing classifier module on the go, I am using optimizer.add_param_group() to add new optimization parameters for the added classifier module and using del optimizer.param_groups[1] to delete the parameters when the module is removed.
> 
> Do you have a minimal example that reproduces the issue?

I will work on this

**Comment by AhmadM-DL** [comment]
> @AhmadM-DL yes, you are right about the cause of the issue. Each `optim.step` call stores/rewrites in the `optimizer.state` optimizer-specific information (exponential average, etc.) about model parameters (these parameters are the keys) so even if you remove a `param_group`, its parameters are still stored in the `optimizer.state` (the next `optim.step` call doesn't touch these parameters and only information about the parameters from other `param_groups` gets upated).
> 
> This line from you bothers me: It runs normally locally but it raises an error on colab or cloud.
> It should be possible to reproduce the error regardless of the environment.
> 
> @agolynski After triage review, maybe I can take this issue and make a PR if there is something to be fixed.

I figured out why there is an inconsistency of output between environments.
In my local pytorchpytorch-cpu 1.1.0] `optimizer.state_dict()` is defined as follow:
```
        # Save ids instead of Tensors
        def pack_group(group):
            packed = {k: v for k, v in group.items() if k != 'params'}
            packed['params'] = [id(p) for p in group['params']]
            return packed
        param_groups = [pack_group(g) for g in self.param_groups]
        # Remap state to use ids as keys
        packed_state = {(id(k) if isinstance(k, torch.Tensor) else k): v
                        for k, v in self.state.items()} # <--------------------------
        return {
            'state': packed_state,
            'param_groups': param_groups,
        }
```
While on CoLab (PyTorch 1.6.0+cu101) ( assumed to be GitHub master branch) is defined as follow:
```        # Save order indices instead of Tensors
        param_mappings = {}
        start_index = 0

        def pack_group(group):
            nonlocal start_index
            packed = {k: v for k, v in group.items() if k != 'params'}
            param_mappings.update({id(p): i for i, p in enumerate(group['params'], start_index)
                                   if id(p) not in param_mappings})
            packed['params'] = [param_mappings[id(p)] for p in group['params']]
            start_index += len(packed['params'])
            return packed
        param_groups = [pack_group(g) for g in self.param_groups]
        # Remap state to use order indices as keys
        packed_state = {(param_mappings[id(k)] if isinstance(k, torch.Tensor) else k): v
                        for k, v in self.state.items()} # <--------------------------
        return {
            'state': packed_state,
            'param_groups': param_groups,
        }
```

If the newest implementation of state_dict to be reserved, I can write a function `remove_param_groups()` to remove the parameter residues in `state`. Actually I will do this to maintain my code.

**Comment by halostorm** [comment]
Pytorch 1.6 use the memory address as the key value. The idmap is assigned when "load_state_dict" from checkpoint. 
So do not change the memory address of param tensors of "optimizer.param_groups". Such as deepcopy().
